### PR TITLE
fix: Specialize typing in `dataframe` between `Frame` and `DataFrame`

### PR DESCRIPTION
**Pull request by luke396**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below.

Seems it should be `IntoDataFrameT`, without changing anything else, just rename it.

### CODE REVIEW

File: narwhals/dataframe.py [hunk 1]
```diff
@@ -33,10 +33,10 @@
     from narwhals.typing import IntoDataFrame
     from narwhals.typing import IntoExpr
 
-FrameT = TypeVar("FrameT", bound="IntoDataFrame")
+IntoDataFrameT = TypeVar("IntoDataFrameT", bound="IntoDataFrame")
```
Comment (MarcoGorelli) [inline narwhals/dataframe.py:36]: we're using `FrameT` for both `DataFrame` and `LazyFrame`

maybe `DataFrame` should use `IntoDataFrameT` and `LazyFrame` should use `IntoFrameT`?

File: narwhals/typing.py [hunk 2]
```diff
@@ -43,7 +43,7 @@ def __dataframe__(self, *args: Any, **kwargs: Any) -> Any: ...
 ]
 """Anything which can be converted to a Narwhals DataFrame or LazyFrame."""
 
-Frame: TypeAlias = Union["DataFrame[Any]", "LazyFrame[Any]"]
+Frame: TypeAlias = Union["IntoFrame", "IntoDataFrame"]
```
Comment (luke396) [inline narwhals/typing.py:46]: Not sure if this is appreciate, or should we create it on top of `dataframe.py`?

Comment (MarcoGorelli) [reply narwhals/typing.py:46]: not sure I follow, sorry, what's the issue with the current definition of `Frame` here to mean "either a Narwhals DataFrame, or a Narwhals LazyFrame"?

**Comment by luke396** [comment]
Notice in #1014, we've addressed the issue I initially wanted to fix. Maybe it should be the answer?
```
-FrameT = TypeVar("FrameT", bound="IntoDataFrame")
+FrameT = TypeVar("FrameT", bound="IntoFrame")
```

**Comment by MarcoGorelli** [comment]
I think it maybe can be specialised still - `FrameT` for `LazyFrame`, and `DataFrameT` for `DataFrame`?

File: narwhals/typing.py [hunk 3]
```diff
@@ -44,12 +44,12 @@ def __dataframe__(self, *args: Any, **kwargs: Any) -> Any: ...
...
 
 Frame: TypeAlias = Union["DataFrame[Any]", "LazyFrame[Any]"]
-"""DataFrame or LazyFrame"""
+"""Narwhals DataFrame or Narwhals LazyFrame"""
 
 # TypeVars for some of the above
 IntoFrameT = TypeVar("IntoFrameT", bound="IntoFrame")
 IntoDataFrameT = TypeVar("IntoDataFrameT", bound="IntoDataFrame")
-FrameT = TypeVar("FrameT", "DataFrame[Any]", "LazyFrame[Any]")
+FrameT = TypeVar("FrameT", bound="Frame")
```
Comment (luke396) [inline narwhals/typing.py:52]: No need, I just noticed it and adjusted to match the others.

**Review by MarcoGorelli — APPROVED** [review]
thanks @luke396 !

**Pull request merged**